### PR TITLE
region_sync:  region syncer support bucket

### LINKED ISSUE #4749: region syncer should support buckets
region syncer can support buckets

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #4749
depences: #4670

### What is changed and how it works?

```commit-message
```

### Check List

Tests

- Unit test

Code changes

- Has persistent data change

Side effects

Related changes

### Release note

```release-note
 None.
```

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -62,6 +62,7 @@ const (
 	// since the once the store is add or remove, we shouldn't return an error even if the store limit is failed to persist.
 	persistLimitRetryTimes = 5
 	persistLimitWaitTime   = 100 * time.Millisecond
+	retry                  = 3
```
Comment (JmPotato) [inline server/cluster/cluster.go:65]: This name is arbitrary, I prefer a more specific one to explain its usage clearly.

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -595,6 +596,27 @@ func (c *RaftCluster) HandleStoreHeartbeat(stats *pdpb.StoreStats) error {
 	return nil
 }
 
+// processBucketHeartbeat update the bucket information.
+func (c *RaftCluster) processBucketHeartbeat(buckets *metapb.Buckets) error {
+	c.RLock()
+	region := c.core.GetRegion(buckets.GetRegionId())
+	c.RUnlock()
+	if region == nil {
+		return errors.Errorf("region %v not found", buckets.GetRegionId())
```
Comment (JmPotato) [inline server/cluster/cluster.go:605]: We should define a new error in `pkg/errs`.

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -595,6 +596,27 @@ func (c *RaftCluster) HandleStoreHeartbeat(stats *pdpb.StoreStats) error {
...
+	}
+	for i := 0; i < retry; i++ {
+		old := region.GetBuckets()
+		// region should not update if the version of the buckets is less than the old one.
+		if old != nil && old.Version >= buckets.Version {
+			return errors.Errorf("the bucket version does not update, region-id:%d ", buckets.GetRegionId())
```
Comment (JmPotato) [inline server/cluster/cluster.go:611]: Ditto.

File: server/cluster/cluster.go (hunk 2 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fix

File: server/cluster/cluster.go (hunk 3 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fix

**Comment by bufferflies** [comment]
PTAL @JmPotato

**Review by JmPotato — APPROVED** [review]

File: server/region_syncer/client.go [hunk 4]
```diff
@@ -186,8 +186,10 @@ func (s *RegionSyncer) StartSyncWithLeader(addr string) {
 				}
 				stats := resp.GetRegionStats()
 				regions := resp.GetRegions()
+				buckets := resp.GetBuckets()
 				regionLeaders := resp.GetRegionLeaders()
 				hasStats := len(stats) == len(regions)
+				hasBuckets := len(buckets) == len(regions)
```
Comment (lhy1024) [inline server/region_syncer/client.go:192]: Is it possible to have more than one bucket in a region?

Comment (bufferflies) [reply server/region_syncer/client.go:192]: It is impossible,  one region has only one buckets properties include many buckets.

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**